File: questions/q530_daily_temperatures/code_double_stack.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<pair<int, int>> store;
        vector<int> ans(temperatures.size(), 0);

        for (int i=temperatures.size()-1; i>=0; i--) {
            int numDays = 1;
            while (!store.empty() && store.top().first <= temperatures[i]) {
                numDays += store.top().second;
                store.pop();
            }

            if (!store.empty()) {
                ans[i] = numDays;
            }

            store.push({temperatures[i], numDays});
        }

        return ans;
    }
};

File: questions/q530_daily_temperatures/code_double_stack.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        ans = [0] * len(temperatures)
        store = []

        for i in range(len(temperatures)-1, -1, -1) :
            numDays = 1
            while len(store)>0 and store[-1][0] <= temperatures[i] :
                numDays += store[-1][1]
                store.pop()
            
            if len(store) > 0 :
                ans[i] = numDays
            
            store.append([temperatures[i], numDays])
        
        return ans
